### DISCUSSION TITLE
3.0.0 beta 4

### DISCUSSION THREAD

**Discussion by zHaytam** [Blazor-Diagrams/Blazor.Diagrams#254] [Announcements]
### Added

- Initial version of Ordering!
	- Nodes, groups and links can now be ordered using the new `Order` property or `SendToFront/Back` methods
	- `Diagram.OrderedSelectables` returns the ordered selectables/models
	- `DiagramCanvas` now uses this new property to render everything
- `GridSnapToCenter` option in order to snap nodes from their center instead of their top/left position (thanks to @[Jeremy Vance](https://github.com/240026763))
- More unit tests

### Changed

- `Groups` is not a list of groups anymore, but a layer instead (just like `Nodes` and `Links`)

### Fixed

- Deleting a group doesn't delete links attached to it
- Deleting a group inside of a group doesn't refresh the parent group
- Links not refreshing when a group's dimensions are updated directly (e.g. deleting a child)
- Layers causing more refreshes than intended

### Removed

- All group-related methods and events from `Diagram`, please use the new layer from now on

<hr /><em>This discussion was created from the release <a href='https://github.com/Blazor-Diagrams/Blazor.Diagrams/releases/tag/3.0.0-beta.4'>3.0.0 beta 4</a>.</em>

**Comment by JeremyGEA** [comment]
Awesome! Thanks for the nod. Happy to help with this very useful library.

**Comment by JeremyGEA** [comment]
I'm having issue with the Beta4 changes for making the default width of user drawn links using the Diagram.Options.Links.Factory.
I changed my code from Beta2 for the link factory to this:
`Diagram.Options.Links.Factory = (diagram, source, targetAnchor) =>
            new LinkModel(new SinglePortAnchor(source as PortModel), targetAnchor)
            {
                Width = LineWidth
            };`
But that crashes the blazor application anytime someone tries to draw a new link from one port to another. What did I do wrong?
I did notice though, that if I update to the latest `develop` branch instead of nuget, it works again after updating the syntax of DefaultRouter and DefaultPathGenerator. I'm guessing a new Beta is coming soon?

**Reply by zHaytam** [reply]
Nothing really changed regarding that feature, so I'm not sure how it's making you crash in beta4 but not using develop.
Do you still have the error?

**Reply by JeremyGEA** [reply]
Here is the error.
![image](https://user-images.githubusercontent.com/49650352/202540965-8f788bef-9e73-4b8b-a9b5-c09a508109d2.png)

I must be doing something wrong because I see this error for the simple demo as well. Just add the follow to the OnInitialized() method in Simple.razor.cs:
```
BlazorDiagram.Options.Links.DefaultRouter = Routers.Orthogonal;
BlazorDiagram.Options.Links.DefaultPathGenerator = PathGenerators.Straight;
BlazorDiagram.Options.Links.EnableSnapping = true;
BlazorDiagram.Options.Links.SnappingRadius = 50;
BlazorDiagram.Options.Links.Factory = (diagram, source, targetAnchor) =>
    new LinkModel(new SinglePortAnchor(source as PortModel), targetAnchor)
    {
        Width = 10
    };
```

**Reply by zHaytam** [reply]
That's weird. You're doing port to port links right? Can you check the type of targetAnchor in the factory?

**Reply by JeremyGEA** [reply]
Yes, I'm doing port to port just like the Simple example.
`targetAnchor` is null. This crash happens right as you click on the port to make a new line. I don't even think a drag event has happened yet.

**Reply by zHaytam** [reply]
Right. I guess I get it now. The OrthogonalRouter doesn't throw an exception anymore, it contains a fallback router now. When you start dragging a new link, the target is not a port so in your version it will throw, but in the new version it will just fallback to a straight router.
